src/imp/mbedtls.rs

```diff
@@ -0,0 +1,575 @@
+            ))
+        }
+    }
+
+    pub fn from_pkcs8(buf: &[u8], key: &[u8]) -> Result<Identity, Error> {
+        let key = Pk::from_private_key(&key.null_terminated(), None)?;
+        let certificates = MbedtlsCert::from_pem_multiple(&buf.null_terminated())?;
+
+        if !certificates.is_empty() {
+            Ok(Identity {
+                key: Arc::new(key),
+                certificates: Arc::new(certificates),
+            })
+        } else {
+            Err(Error::Custom(
+                "X509 chain file is missing certificate chain".to_owned(),
+            ))
+        }
+    }
+
+    fn certificates(&self) -> Arc<MbedtlsList<MbedtlsCert>> {
+        self.certificates.clone()
+    }
+
+    fn private_key(&self) -> Arc<Pk> {
+        self.key.clone()
+    }
+}
+
+impl Debug for Identity {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Identity")
+            .field(
+                "certificates",
+                &self
+                    .certificates
+                    .iter()
+                    .map(|cert| cert.as_der().to_vec())
+                    .collect::<Vec<_>>(),
+            )
+            .field(
+                "key_name",
+                &self.key.name().map(String::from).map_err(Error::Tls),
+            )
+            .finish()
+    }
+}
+
+#[derive(Clone)]
+pub struct Certificate(MbedtlsBox<MbedtlsCert>);
+
+impl Certificate {
+    pub fn from_der(buf: &[u8]) -> Result<Certificate, Error> {
+        let cert = MbedtlsCert::from_der(buf).map_err(Error::Tls)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn from_pem(buf: &[u8]) -> Result<Certificate, Error> {
+        let cert = MbedtlsCert::from_pem(&buf.null_terminated()).map_err(Error::Tls)?;
+        Ok(Certificate(cert))
+    }
+
+    pub fn to_der(&self) -> Result<Vec<u8>, Error> {
+        let der = self.0.as_der().to_vec();
+        Ok(der)
+    }
+}
+
+impl Debug for Certificate {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Certificate")
+            .field(&self.0.as_der())
+            .finish()
+    }
+}
+
+pub struct TlsStream<S: Stream> {
+    ctx: Arc<Mutex<Context>>,
+    role: Endpoint,
+    identity: Option<Identity>,
+    _stream: PhantomData<S>,
+}
+
+impl<S: Stream> TlsStream<S> {
+    pub fn get_ref(&self) -> impl Deref<Target = S> + '_ {
+        // Is this unsound?
+        // map requires FnOnce(&mut T) -> &mut U, so I use io_mut()
+        // there. Return type is Deref, not DerefMut, so does it break
+        // the only single mutable reference rule?
+        // But this shouldn't be possible without unsafe
+        MutexGuard::map(self.ctx.lock(), |ctx| {
+            ctx.io_mut()
+                .expect("Not connected")
+                .downcast_mut::<S>()
+                // unwrap is safe as this is the exact stream passed to context
+                .unwrap()
+        })
+    }
```

Comment:
Calling this function now may easily deadlock, if any other method (eg. this itself) that locks the context is called while holding the reference. this could be a `ReentrantMutex` instead, but not yet sure if we can get `get_mut` working then (lock doesn't provide `&mut`). Will try with `RefCell` in the meantime but again, it's not that simple :sweat_smile:

Reply:
no longer relevant